feat: typed Inputs and Outputs

J-Loudet opened the pull request:
This PR adds the possibility for users to associate a type to the Inputs and
Outputs of their Nodes.

Associating a type has two benefits:
1. On an Output, Rust will enforce that only an instance of the associated type
   is sent to downstream nodes,
2. On an Input, Zenoh-Flow will, transparently, try to "convert" the received
   data to a reference of the associated type.
   
Raw versions of the Input and Output, namely `InputRaw` and `OutputRaw` are
available to have non-typed variant.

The method `forward` was added to the `OutputRaw` for cases that require
performance: Zenoh-Flow will perform no verification and just forward the a
`LinkMessage` on the Output.

## Internal modifications

The `Inputs` and `Outputs` type aliases were rewritten in favor of concrete
structures. This allowed removing the (badly named) `Streams` traits.

The `Message` structure was renamed `LinkMessage`: it better describes how
it's used and where it "comes from".

The `Data` structure was renamed `Payload`.

The purpose of the last two modifications is to be able to present users with a
`Message` enumeration: it is either `Data<T>` or a `Watermark`.

Lastly, the synchronous methods `send` and `recv` were removed.

## Examples

``` rust
// `outputs` is an `Outputs` structure.
let output: Output<usize> = outputs.take("out").unwrap();
// alternative:
// let output = outputs.take::<usize>("out").unwrap();

// `inputs` is an `Inputs` structure.
let input: Input<usize> inputs.take("in").unwrap();
// alternative:
// let input = inputs.take::<usize>("out").unwrap();
```

gabrik committed:
WIP: initial re-refacto of node traits

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
WIP: removed old traits, added docs for new traits, updated dataflow test, cleanup

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
update macros

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
updated loader and macros

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
Addressing comments

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
Fix typo

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
reorganized Source, Sink, Operator and Node traits

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

gabrik committed:
hidden Factory, new derive macros for export (added to the prelude)

Signed-off-by: gabrik <gabriele.baldoni@gmail.com>

Commit:
refacto: remove Callback shorthand

Supporting the "transformation" of Input / Output into Callback introduced
constraints in our API (and internal codebase). This PR removes this support to
simplify our API.

One of the main benefit is the signature of the "constructors":

before:
```rust
fn new(
   context: &mut Context,
   configuration: &Option<Configuration>,
   mut outputs: Outputs
   ) -> Result<Option<Self>>;
```

after:
```rust
fn new(
   context: &mut Context,
   configuration: &Option<Configuration>,
   mut outputs: Outputs
   ) -> Result<Self>;
```

Commit:
refacto: replace internal Factory by constructors

By internally mapping the `new` static method of each Node to a "constructor"
function (e.g. `SourceFn`) we remove the need of the additional trait `Factory`.

To make each constructor asynchronous while not staining the entire API with
lifetime annotations, the different Node traits have been modified to take no
reference: `&mut Context` became `Context` and `&Option<Configuration>` became
`Option<Configuration>`.

Commit:
feat: (wip) Input and Output are now typed

This commit makes the `Input` and `Output` structure typed. The first change is
what Nodes store:

(before)
```rust
struct MyOperator {
    input_1: Input,
    input_2: Input,
    output: Output,
}
```

(after)
```rust
// T, U, V are 3 distinct types that implement ZFData
struct MyOperator {
    input_1: Input<T>,
    input_2: Input<U>,
    output: Output<V>,
}
```

The second change is that users receive a `Data<T>` after a call to
`input.recv_async().await`. As `Data<T>` dereferences to `&T` there is no longer
the need to call `try_get::<T>()`.

Similarly, an `Output` expects a `T` for its `send_async` method.

TODO:
- fix macros
- clean up code
- add comments
- test python bindings

---

- [x] I have successfully launched a flow (single daemon)

Commit:
feat: implement `ZFData` for basic Rust types

Commit:
Merge branch 'feat/zfdata-basic-types' into feat/typed-Inputs-Outputs

Commit:
refacto: `send_async` accepts `impl Into<Data<T>>`

The purpose of this change is to enable passing `Data<T>` directly, without
having to clone the underlying `T`.

This should lead to better performance.

Commit:
refacto: split `io.rs` and move to separate folder

Commit:
Merge branch 'dev/v0.4.0' into feat/typed-Inputs-Outputs

Commit:
doc: document Input + Output structures and methods

Commit:
update documentation

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+/// `Input<T>` conveniently receives instances of `T` and is thus less performant as Zenoh-Flow has
+/// to manipulate the data; an `InputRaw` is more performant but all the manipulation must be
+/// performed in the code of the Node. An `InputRaw` can be leveraged, for instance, for
+/// load-balancing or rate-limiting.
```
I would also add that `InputRaw` is used by bindings. As Python (or other bindings in the future) cannot understand types coming from Rust (and vice-versa)

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+    ///
+    /// ## Performance
+    ///
+    /// With a typed [`Input<T>`](`Input`), Zenoh-Flow will perform operations on the underlying
```
We could add that deserialization is done only if the inner data is serialized (i.e., coming from the network or the upstream node sent it as a byte array).

J-Loudet replied:
The text in the line after isn't explicit enough?

> if the data are received serialized then Zenoh-Flow will deserialize them, if they are received "typed" then Zenoh-Flow will check that the type matches what is expected.

gabrik replied:
I missed it :)

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+/// It's primary purpose is to ensure "optimal" performance. This can be useful to implement
+/// behaviour where actual access to the underlying data is irrelevant.
+#[derive(Clone, Debug)]
+pub struct InputRaw {
```
What happens if I use an `Output<T>` to send data to an `InputRaw`? 
My guess is that data need to be serialized by the `InputRaw::recv_async`, if not I guess that something bad will happen.

J-Loudet replied:
Nothing bad happens! :-)

An `Output<T>` will send a `Payload::Typed` to downstream nodes. Zenoh-Flow will thus serialize it (as `T` implements `ZFData`) only if need be.

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+    /// ## Error
+    ///
+    /// If an error occurs on one of the channels, this error is returned.
+    pub async fn recv_async(&self) -> Result<LinkMessage> {
```
Are we going to add a `try_recv()` as briefly discussed?

J-Loudet replied:
I did not want to "mix" too many changes in the same PR. As we want to remove the, potentially harmful, synchronous `recv` I saw no point of updating it to the changes and so removed it now. But this leaves things in an "half-done" state, so I'll add the `try_{recv|send}` and rename the methods `{recv|send}_async` to `{recv|send}`. :)

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+    /// Several errors can occur:
+    /// - one of the channels can return an error (e.g. a disconnection),
+    /// - Zenoh-Flow failed at interpreting the received data as an instance of `T`.
+    pub async fn recv_async(&self) -> Result<(Message<T>, Timestamp)> {
```
Are we going to add a `try_recv()` as briefly discussed?

gabrik replied:
The PR description says that this function name should be `recv`.

gabrik commented on zenoh-flow/src/io/output.rs:
```diff
@@ -0,0 +1,280 @@
+    /// If an error occurs while sending the watermark on a channel, Zenoh-Flow still tries to send
+    /// it on the remaining channels. For each failing channel, an error is logged and counted for.
+    /// The total number of encountered errors is returned.
+    pub async fn send_async(&self, data: impl Into<Payload>, timestamp: Option<u64>) -> Result<()> {
```
Need a `try_send`, similar to `try_recv`

gabrik replied:
The PR description says that this function name should be `send`.

gabrik commented on zenoh-flow/src/io/output.rs:
```diff
@@ -0,0 +1,280 @@
+    /// If an error occurs while sending the message on a channel, Zenoh-Flow still tries to send it
+    /// on the remaining channels. For each failing channel, an error is logged and counted for. The
+    /// total number of encountered errors is returned.
+    pub async fn send_async(&self, data: impl Into<Data<T>>, timestamp: Option<u64>) -> Result<()> {
```
Same as above

gabrik replied:
The PR description says that this function name should be `send`.

gabrik commented on zenoh-flow/src/io/input.rs:
```diff
@@ -0,0 +1,190 @@
+    /// ## Error
+    ///
+    /// If an error occurs on one of the channels, this error is returned.
+    pub async fn recv_async(&self) -> Result<LinkMessage> {
```
The PR description says that this function name should be `recv`.

gabrik commented on zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:
```diff
@@ -30,7 +30,7 @@ use zenoh_util::core::AsyncResolve;
 pub(crate) struct ZenohSender {
     pub(crate) id: NodeId,
-    pub(crate) input: Input,
+    pub(crate) receivers: Vec<flume::Receiver<LinkMessage>>,
```
Why is not using the `Input`?

J-Loudet replied:
It's an oversight, I introduced the `InputRaw` and `OutputRaw` after making that change & forgot to port it

gabrik commented on zenoh-flow/src/runtime/dataflow/instance/runners/connector.rs:
```diff
@@ -87,26 +87,33 @@ impl Node for ZenohSender {
 /// A `ZenohReceiver` receives the messages from Zenoh when nodes are running on different runtimes.
 pub(crate) struct ZenohReceiver {
     pub(crate) id: NodeId,
-    pub(crate) output: Output,
```
Why it is not using the `RawOutput`?

gabrik commented on zenoh-flow/src/types/message.rs:
```diff
@@ -183,22 +314,169 @@ impl Message {
-impl Eq for Message {}
+impl Eq for LinkMessage {}
+
+/// A `Message<T>` is what is received on an `Input<T>`, typically after a call to `recv_async` or
+/// `recv`.
```
`recv` or `try_recv`

gabrik commented on zenoh-flow/src/types/message.rs:
```diff
@@ -183,22 +314,169 @@ impl Message {
+    // CAVEAT: Setting the value **typed** to `None` when it was already set to `Some` can cause a
+    // panic! The visibility of the entire structure is private to force us not to make such
+    // mistake.
+    // In short, tread carefully!
```
Typo: treat

J-Loudet replied:
Although "treat" would indeed work, I did want to use the expression "tread carefully" (= be cautious) 😛

gabrik commented on zenoh-flow/src/types/message.rs:
```diff
@@ -183,22 +314,169 @@ impl Message {
+}
+
+// CAVEAT: The implementation of `Deref` is what allows users to transparently manipulate the type
+// `T`. However, for things to go well, the `typed` field MUST BE MANIPULATED WITH CAUTION.
```
Add a Warning to the doc, not only the caps text

gabrik commented on zenoh-flow/tests/dataflow.rs:
```diff
@@ -105,8 +107,8 @@ impl Node for GenericSink {
 // OPERATORS
 
 struct NoOp {
```
I would add an Operator with `Input` and `RawOutput` (and vice-versa) as well as a `Sink` with `RawInput` just to verify that all cases work.

gabrik reviewed (changes requested):
Overall seems good, I guess there are still some cases that may cause troubles, as well as some things that need to be changed in function names (according to PR description) and in the `ZenohSender`/`ZenohReceiver` that should not manipulate directly the `flume::channel` but use the `Input`/`Output`.

Commit:
refacto: incorporate feedback

- updated comments where appropriate

- in input.rs:
  - renamed `recv_async` to `recv`
  - added synchronous but non-blocking `try_recv`
  - both `recv` and `try_recv` behave similarly: they only error out when all
    the channels are disconnected

- in output.rs:
  - renamed `send_async` to `send`
  - added synchronous but non-blocking `try_send`
  - renamed `send_to_all_async` to `forward` (which allowed me to remove few
    lines of code)

- tests/dataflow.rs: was refactored to test all possible "exchange" of data.
  From Raw to Typed, the other way around, from an Output to Input, the other
  way around and mixing everything

gabrik reviewed (approved):
LGTM!
